### PR TITLE
add standard for integration and e2e tests

### PR DESCRIPTION
**Pull request by jdkandersson**
Adds a standard for when to write and what to cover in integration tests as a follow up from this conversation: https://github.com/canonical/indico-operator/pull/220#discussion_r1091363449

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -317,6 +318,58 @@ Alternatives to merging the PR with failing status checks include:
...
+or tool. The following examples illustrate what this means by providing guidance
+on what should and should not be covered by integration/ e2e tests.
+
+The following should be covered by integration/ e2e tests:
+
+* An action the charm provides
+* An integration the charm provides
+* A configuration the charm provides
+* That the workload is up and running
+* The features of a tool
```
Comment (merkata) [inline CONTRIBUTING.md:345]: Can this be expanded or rephrased? What is a `tool` in the scope of a charm?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -317,6 +318,58 @@ Alternatives to merging the PR with failing status checks include:
...
 
+## When to Write and What to Cover In Integration/ End-To-End Tests
+
+The team creates charms and tooling used and intended to be used directly or as
+building blocks for mission critical purposes by both Canonical and external
+users. The charms and tools we provide need to meet a high quality bar to ensure
+that they work for our users as intended. If we ship charms and tools that don't
+meet this high standard, the impact could be widespread.
+
+This standards covers when to write integration/ end-to-end (e2e) tests and also
```
Comment (arturo-seijas) [inline CONTRIBUTING.md:329]: s/standards/standard

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -317,6 +318,58 @@ Alternatives to merging the PR with failing status checks include:
...
+
+The reason these examples do not have to be tested is because we are not
+accountable for them and their owners need to ensure they keep working.
+
+To address the second concept of `works as advertised`, when writing an
+integration/ e2e test, it is not sufficient to just check that, for example,
+that Juju reports that running the action was succesfull. Additional checks need
+to be executed to ensure that whatever the action was intended to achieve
+worked. An example where an exception could apply is if it is unduly expensive
+or impractical to perform that check.
```
Comment (gregory-schiano) [inline CONTRIBUTING.md:366]: That's the main point of discussion in the indico PR
Let's wait to close this discussion before merging this.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -317,6 +318,58 @@ Alternatives to merging the PR with failing status checks include:
...
+usually need to be covered by integration/ e2e tests:
+
+* A feature provided by the workload which is not enhanced by the charm
+* That GitHub works
+* That the network works
+* That the operating system works
+* That Kubernetes works
+* That Juju works
+
+The reason these examples do not have to be tested is because we are not
```
Comment (amandahla) [inline CONTRIBUTING.md:358]: Could this be rephrased or even removed since is implicit in the previous sentence?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -317,6 +318,58 @@ Alternatives to merging the PR with failing status checks include:
 
 This will ensure that we minimise the number of bugs in our code and tooling.
 
+## When to Write and What to Cover In Integration/ End-To-End Tests
```
[the remark this answers is not in the document]
Comment (jdkandersson) [reply CONTRIBUTING.md:321]: Interesting perspective. The reason e2e is included is because of things like triggering action on a charm seems to be an e2e test. The `upload-charm-docs` also includes e2e tests that update docs on discourse and perform a documentation migration. Do you think those qualify as e2e tests? Are there scenarios that should be covered which are e2e which aren't currently covered?

File: CONTRIBUTING.md (hunk 3 above)
Comment (jdkandersson) [reply CONTRIBUTING.md:366]: As discussed yesterday, it would be good if we could have a few alternatives to consider for this statement, e.g., a sentence that gives more discretion

File: CONTRIBUTING.md (hunk 4 above)
Comment (jdkandersson) [reply CONTRIBUTING.md:358]: It is useful to give reasons in standards that are explicit so that people know why we standard is in place. Any rephrasing that you have in mind?

File: CONTRIBUTING.md (hunk 1 above)
Comment (jdkandersson) [reply CONTRIBUTING.md:346]: added a definition of what a tool is earlier. It is intended to be things like a GitHub action

File: CONTRIBUTING.md (hunk 2 above)
Comment (jdkandersson) [reply CONTRIBUTING.md]: Thanks, fixed

File: CONTRIBUTING.md (hunk 3 above)
Comment (arturo-seijas) [reply CONTRIBUTING.md:367]: I wouldn't consider exceptions here, but I wouldn't put a third party under test either. The minimal requirement here would be to have proper unit tests. Integration tests should be delivered if it's possible to mock the third party component. If not, complexity and criticality should be the criteria to consider.

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply CONTRIBUTING.md:321]: Yes, those certainly (or in the action case, might be) are but I would consider them a different kind of animal. E2E tests should ideally be kept to a minimum given they are slower, potentially more flaky and do not pinpoint the problematic components/code in case of an error.
I'd rather include a separate section for E2E tests to describe in which conditions are applicable: when no integration tests are possible, feasible or the functionality is critical. I suggest we leave those out of the scope of this proposal for now, since we might want to think about smoke tests too. After all, we are mainly working with distributed systems and smoke test might also reduce the need for E2E testing.

File: CONTRIBUTING.md (hunk 4 above)
Comment (amandahla) [reply CONTRIBUTING.md:359]: This is my suggestion for your consideration:
```
The following examples illustrate what this means by guiding what should and should not be covered by integration/ e2e tests.

These should be covered because we are accountable for these features of a charm or tool by being its owner:

- An action the charm provides
- An integration the charm provides
- A configuration the charm provides
- That the workload is up and running
- The features of a tool

And these do not usually need to be covered because we are not accountable for them, and their owners need to ensure they keep working:

- A feature provided by the workload which is not enhanced by the charm
- That GitHub works
- That the network works
- That the operating system works
- That Kubernetes works
- That Juju works
```

BTW "do not usually"...What would be the exceptions?

File: CONTRIBUTING.md (hunk 3 above)
Comment (jdkandersson) [reply CONTRIBUTING.md:367]: One of the key reasons for adding this section was to indicate that integration tests should check that the integration with an external dependency works. So, for example, that an action that adds a user is executed and then the success of running that action is check without mocking. Without that kind of test, how would we know whether the action that was added works?

File: CONTRIBUTING.md (hunk 4 above)
Comment (jdkandersson) [reply CONTRIBUTING.md:359]: I can't think of any, just leaving it open if something comes up in the future

File: CONTRIBUTING.md (hunk 3 above)
Comment (merkata) [reply CONTRIBUTING.md:367]: This is a good discussion. I think we are marching into a grey area however. I think the intent of third parties is that we use reputable and well maintained sources. We have dedicated a whole paragraph to it at https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#dependencies and to me there is a reason behind it, although it is not written there - the more stable, well-maintained and reputable the dependency, the more trustworthy the contract it offers. A contract is the binding between the dependency and our code and this establishes the level of trust/confidence when using it. And that level of trust is what we're after, so we don't have to test events outside of our code base. 

To me an API that should create a user and returns a response according to the state of the action is the thing I want to test. I believe in that contract and that the third party library will deliver based on that. That's how it is in real life as well. I can choose to distrust that official contract (the API) and do all sorts of background checks - when the API is called, does it create a field in a DB, what if there's no connection to the DB, can a regular user do some admin action that they're not supposed to do, does an admin have all the documented privileges they're supposed to have, what happens if I delete and create the same user etc.? Having a stable API gives you the comfort of testing S3 without using an actual S3 and calling the API that is implemented in `localstack` for instance. That's an improvement imo. 

So to recap, just because we can, doesn't mean we should. To me an integration test should verify the contractual clause (the integration point) that is in question - going further than that negates the benefits of having a contract in the first place.

Comment (jdkandersson) [reply CONTRIBUTING.md:367]: Yes, I don't think we should re-test our dependency and the proposal isn't intended to imply that. Which part is suggesting re-testing the dependency?

The integration test should not be designed as tests for the dependency. Instead, they should be designed to check that our code properly interacts with the dependency in an environment as close as possible to production. I do agree that there is room for judgement here in terms of the extend to which these kind of tests are created. Arturo's feedback on business criticality is a good point as well. I'm curios on any proposals for alternate wording.

One of the key challenges is the extend of checking whether the integration worked. I was trying to use GitHub script to add team reviewers to a PR yesterday. I called the API according to its contract and it did not throw an error. That means it worked, right? No it didn't - there was a permission problem that was silently being ignored and the call didn't end up doing anything. Is the dependency poor? We probably would accept it as it is well maintained. So if we had this kind of a case for one of our charms, just relying on the response from the dependency for checks isn't always sufficient. It might be in some cases, in others it might not.

Another example was for the `upload-charm-docs` migration feature. The tests that were written used the `git` contract correctly and everything seemed to be working. However, when the action was deployed to production, it was broken. Why? The applicable contract for how to use `git` was different when the action was running in production than was assumed in the tests. How would we guard against that without running the code in a production like environment and check that what we expected to happen did occur?

For operational code, reliability is at a premium. You would expect SREs to prioritise reliability, limit regressions and maximise uptime (depending on the business criticality of an application). If we are automating operations, we should also have similar high standards.

Comment (gregory-schiano) [reply CONTRIBUTING.md:367]: After discussing here and there I think that what could make consensus is to think about business criticality. 

But that's a tough topic when it comes to defining it feature per feature as its subjective, and needs to be defined everytime a new feature is developed.

My proposal is to take this from a more global perspective and define business criticality per charm. We'll then have 2 types of charms:
- Business Critical charm (e.g. Jenkins): We do extensive integration tests there, that could even be seen as "not trusting the contract" as a non detected issue on this charm can have big business impact. There could even be some more rules that applies to this type of charms (like building up to date docker images every day, running integration tests every day, scanning for CVE every day, ...)
- Non Business Critical Charm: Covering these charms with extensive integration tests is not mandatory, and it will be decided on a case by case basis. Meaning that it'll be defined either during estimation, or spec definition, or at last during code review. But we don't expect these charms to have integration tests that don't trust the contract. This can for sure have impact, contracts are sometimes blurry or could vary depending on the environment, but we estimate that this won't have any critical business impact and that the team can be reactive enough to fix these issues quickly.

One point we should align also is the definition of: Integration tests, unit tests, e2e tests and smoke tests
Proposal to be refined:
- Unit tests
These are tests that are covering charm / service methods to ensure that given a specific context, mocked interfaces, the method is giving the expected output. Tests could cover more than one method if some methods aren't doing business logic, but these tests shouldn't be functional, they just ensure that the code is doing what it's supposed to do. A test that requires too many mocks is the sign of lack of design, and requires refactoring
- Integration tests
These are test that ensure that the charm integrated with its dependencies will behave properly. It doesn't test the code in a production like environment, meaning that we don't connect to ressources equivalent environment or environment with existing datasets. The integration tests will spawn specific dependencies that could differ from the production one (e.g we don't run on real openstack, but localstack, we don't run on real DB cluster but a single node charm, ...). These tests are functional, they ensure that the features provided by the charm are working as intended, but they don't necessarily need to ensure that the API/Service/CLI they're interacting with are working as intended.
- E2E tests
These are simulated user scenarios running on an environment as close as possible to production. These are usual ran on a staging environment with preexisting data, and interacting as the user would do (ideally using juju cli). These are functional tests and ensure that the charm is working as intended in the condition close to production in order to detect issues related to this environment (ressources, pre-existing condition, migrations, ...)
- Smoke tests
These are functional tests focusing on a business critical path used to ensure that the most critical features are still working after a production deployment. The aim is to ensure that a deployment didn't impact business continuity and detect potential defects immediately after a production release.

All this probably requires some refinement, feel free to adapt :)

Comment (jdkandersson) [reply CONTRIBUTING.md:367]: Thanks for this, I think this is reasonable. I would add that, if we see a bug that we needed to fix, a test that catches the bug is first added and then the bug is fixed. This means that the default for a non-business critical charm might be that no e2e tests are required for a given feature. If we find a bug that can only be caught using an e2e test - it means an e2e test should now be considered when it might have been skipped in the past.

I'll redraft this PR to incorporate the feedback

File: CONTRIBUTING.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jdkandersson) [reply CONTRIBUTING.md]: Sounds good, let's focus on integration tests here. Removed e2e tests

File: CONTRIBUTING.md (hunk 3 above)
Comment (jdkandersson) [reply CONTRIBUTING.md]: Updated the standard to incorporate this, I'll close this conversation for now

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -317,6 +369,62 @@ Alternatives to merging the PR with failing status checks include:
...
+charms:
+
+* An action the charm provides
+* An integration the charm provides
+* A configuration the charm provides
+* That the workload is up and running
+* The features of a tool
+
+The reason these examples should be covered is because we are accountable for
+these features of a charm or tool by being it's owner. Charms that are not
```
Comment (amandahla) [inline CONTRIBUTING.md:401]: it's -> its

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -317,6 +369,62 @@ Alternatives to merging the PR with failing status checks include:
...
+* That GitHub works
+* That the network works
+* That the operating system works
+* That Kubernetes works
+* That Juju works
+
+To address the second concept of `works as advertised`, when writing an
+integration test for business critical charms, it is not sufficient to just
+check that, for example, that Juju reports that running the action was
+succesfull. Additional checks need to be executed to ensure that whatever the
```
Comment (amandahla) [inline CONTRIBUTING.md:418]: succesfull -> successful

**Review by gregory-schiano — DISMISSED** [review]
LGTM, thx for the proposal David

File: CONTRIBUTING.md (hunk 6 above)
Comment (jdkandersson) [reply CONTRIBUTING.md]: thanks, fixed

File: CONTRIBUTING.md (hunk 7 above)
Comment (jdkandersson) [reply CONTRIBUTING.md]: thanks, fixed

**Review by yanksyoon — APPROVED** [review]
LGTM :)

**Review by arturo-seijas — APPROVED** [review]
LGTM
